README.md
# Memorable Experience - Logging and performances

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Setup and configuration](#setup-and-configuration)
  - [Required Python Packages](#required-python-packages)
  - [Elasticsearch Migrator](#elasticsearch-migrator)
  - [Environment variables](#environment-variables)
    - [Web Service](#web-service)
    - [Script main](#script-main)
    - [Script id_email](#script-id_email)
    - [Script apps_usage](#script-apps_usage)
- [Usage](#usage)
  - [Web service](#web-service)
  - [Script for computing the analytics](#script-for-computing-the-analytics)
  - [Script for getting the associations](#script-for-getting-the-associations)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

There is a web service. It can be used to store, retrieve and delete messages (+ logs and performances, but not sure they work as documented or as should work).

The web service need an Elasticsearch database for storing data.

It could be used a Celery (a Distributed Task Queue) instance for handling the post request of analytics.
It requires a broker and a backend. An example to use it, having a Redis instance on the `localhost` on the port `6379` in db number `0`, is by setting the two env vars in this way:

```python
CELERY_BROKER_URL='redis://localhost:6379/0'
CELERY_RESULT_BACKEND='redis://localhost:6379/0'
```

You'll need one `celery beat` scheduler and at least one `celery worker` to get things done. Run the following command in a separate terminal tab to run the celery beat:

```bash
celery beat -A memex_logging.celery.initialize.celery
```

You can run the following command as many times you want in order to run several workers:

```bash
celery worker -A memex_logging.celery.initialize.celery
```

Notes:
- Celery requires all the environment variables of the web service to work. 
- If you run the celery beat using the provided docker images, you need to set a volume un the `/celery` directory of the container. Otherwise, the container may fail to start


There is a `main` script for computing the analytics, questions, users. Analytics, questions and users will be stored in dedicated .csv/.tsv files.
It also extracts the tasks, the profiles and a dump of the messages using the logger, the task manager, the hub, the profile manager and the incentive server. The tasks, the profiles and the messages will be stored in dedicated .json files.

There is a `id_email` script for getting the association between the id and the email of users. This will be stored in a dedicated .csv/.tsv file.

Finally, there is a `apps_usage` script for getting the app usage of the users and the dump of the profiles without personal information.
The user app usage will be stored in a dedicated .csv/.tsv file. The dump of the profiles without personal information will be stored in dedicated .json file.


## Setup and configuration

### Required Python Packages

Required Python packages can be installed using the command:

```bash
pip install -r requirements.txt
```


### Elasticsearch Migrator

Before starting the services, apply migration to your Elasticsearch instance.
To run the migration you only need to run the migration module:

```bash
python -m memex_logging.migration.migration
```

In order to execute the migrations, configure the following environment variables:

* `MIGRATION_FOLDER`  (optional, the default value is `actions`): the folder which contains all the migration;
* `MIGRATION_MANAGER_INDEX`  (optional, the default value is `migrations`): the index in which the applied migrations are stored;
* `EL_HOST` (optional, the default value is `localhost`): the host where the Elasticsearch database is going to be available;
* `EL_PORT` (optional, the default value is `9200`): the port where the Elasticsearch database is going to be available;
* `EL_USERNAME` (optional for versions of Elasticsearch < `7`): the username of the user to access the Elasticsearch database;
* `EL_PASSWORD` (optional for versions of Elasticsearch < `7`): the password of the user to access the Elasticsearch database;


### Environment variables

#### Web Service

The web service allows to set the following environment variables:

* `WS_HOST` (optional, the default value is `0.0.0.0`): the host where the web service is going to be available;
* `WS_PORT` (optional, the default value is `80`): the port where the web service is going to be available;
* `EL_HOST` (optional, the default value is `localhost`): the host where the Elasticsearch database is going to be available;
* `EL_PORT` (optional, the default value is `9200`): the port where the Elasticsearch database is going to be available;
* `EL_USERNAME` (optional for versions of Elasticsearch < `7`): the username of the user to access the Elasticsearch database;
* `EL_PASSWORD` (optional for versions of Elasticsearch < `7`): the password of the user to access the Elasticsearch database;
* `INSTANCE`: the host of target instance;
* `APIKEY`: the apikey for accessing the services;
* `CELERY_BROKER_URL`: the information about the broker to use the Celery instance, it must be in the following format: `redis://:password@hostname:port/db_number`;
* `CELERY_RESULT_BACKEND`: the information about the result backend to use the Celery instance, it must be in the following format: `redis://:password@hostname:port/db_number`.
* `CARDINALITY_PRECISION_THRESHOLD` (optional, the default value is `40000`): the precision threshold parameter for cardinality aggregations (maximum supported value is `40000`).

Optionally is it possible to configure sentry in order to track any problem. Just set the following environment variables:

* `SENTRY_DSN` (optional) The data source name for sentry, if not set the project will not create any event
* `SENTRY_RELEASE` (optional) If set, sentry will associate the events to the given release
* `SENTRY_ENVIRONMENT` (optional) If set, sentry will associate the events to the given environment (ex. `production`, `staging`)


#### Script main

The script for computing the analytics, questions, users and for extracting the tasks, the profiles and the messages allows to set the following environment variables:

* `ANALYTIC_FILE`: the path of the csv/tsv file where to store the analytics, it can also be set using the argument `-af` or `--analytic_file` when manually running the Python service;
* `QUESTION_FILE`: the path of the csv/tsv file where to store the questions, it can also be set using the argument `-qf` or `--question_file` when manually running the Python service;
* `USER_FILE`: the path of the csv/tsv file where to store the users, it can also be set using the argument `-uf` or `--user_file` when manually running the Python service;
* `PROFILE_FILE`: the path of the json file where to store user profiles, it can also be set using the argument `-of` or `--profile_file` when manually running the Python service;
* `TASK_FILE`: the path of the json file where to store the tasks, it can also be set using the argument `-tf` or `--task_file` when manually running the Python service;
* `MESSAGE_FILE`: the path of the json file where to store the dump of messages, it can also be set using the argument `-mf` or `--message_file` when manually running the Python service;
* `INSTANCE` (optional, the default value is the development instance): the host of target instance, it can also be set using the argument `-i` or `--instance` when manually running the Python service;
* `APIKEY`: the apikey for accessing the services, it can also be set using the argument `-a` or `--apikey` when manually running the Python service;
* `APP_ID`: the id of the application in which compute the analytics, it can also be set using the argument `-ai` or `--app_id` when manually running the Python service;
* `ILOG_ID`: the id of the ilog application to check if the user has enabled it or not, it can also be set using the argument `-ii` or `--ilog_id` when manually running the Python service;
* `SURVEY_ID`: the id of the survey application to check if the user has enabled it or not, it can also be set using the argument `-si` or `--survey_id` when manually running the Python service;
* `TIME_RANGE`: the temporal range in which compute the analytics, example of allowed values are ["TODAY", "1D", "7D", "10D", "30D"], it can also be set using the argument `-r` or `--range` when manually running the Python service.

Alternatively to the `TIME_RANGE` arbitrary start and end time could be set using the following environment variables:

* `START_TIME`: the start time from which compute the analytics, must be in iso format, it can also be set using the argument `-s` or `--start` when manually running the Python service;
* `END_TIME`: the end time up to which compute the analytics, must be in iso format, it can also be set using the argument `-e` or `--end` when manually running the Python service.


#### Script id_email

The script for getting the association between the id and the email of users allows to set the following environment variables:

* `FILE`: the path of the csv/tsv file where to store the id-email associations, it can also be set using the argument `-f` or `--file` when manually running the Python service;
* `INSTANCE` (optional, the default value is the development instance): the host of target instance, it can also be set using the argument `-i` or `--instance` when manually running the Python service;
* `APIKEY`: the apikey for accessing the services, it can also be set using the argument `-a` or `--apikey` when manually running the Python service;
* `APP_ID`: the id of the application in which compute the analytics, it can also be set using the argument `-ai` or `--app_id` when manually running the Python service;
* `TIME_RANGE` (optional): the temporal range in which compute the analytics, example of allowed values are ["TODAY", "1D", "7D", "10D", "30D"], it can also be set using the argument `-r` or `--range` when manually running the Python service.

Alternatively to the `TIME_RANGE` arbitrary start and end time could be set using the following environment variables:

* `START_TIME`: the start time from which compute the analytics, must be in iso format, it can also be set using the argument `-s` or `--start` when manually running the Python service;
* `END_TIME`: the end time up to which compute the analytics, must be in iso format, it can also be set using the argument `-e` or `--end` when manually running the Python service.


#### Script apps_usage

The script for getting the app usage of the users, the dump of the profiles without personal information and profiles and task in the new format allows to set the following environment variables:

* `USER_FILE`: the path of csv/tsv file where to store the users apps usage, it can also be set using the argument `-uf` or `--user_file` when manually running the Python service;
* `PROFILE_FILE`: the path of json file where to store the profiles without personal information, it can also be set using the argument `-pf` or `--profile_file` when manually running the Python service;
* `INSTANCE` (optional, the default value is the development instance): the host of target instance, it can also be set using the argument `-i` or `--instance` when manually running the Python service;
* `APIKEY`: the apikey for accessing the services, it can also be set using the argument `-a` or `--apikey` when manually running the Python service;
* `APP_IDS`: the ids of the chatbots from which take the users. The ids should be separated by `;`, it can also be set using the argument `-ai` or `--app_ids` when manually running the Python service;
* `ILOG_ID`: the id of the ilog application to check if the user has enabled it or not, it can also be set using the argument `-ii` or `--ilog_id` when manually running the Python service;
* `SURVEY_IDS`: the ids of the survey applications to check if the user has enabled it or not. The ids should be separated by `;`, it can also be set using the argument `-si` or `--survey_ids` when manually running the Python service;
* `USER_UPDATES_DUMPS`: the paths of csv/tsv files with the dump of the user that updated the profile. The paths should be separated by `;`, it can also be set using the argument `-u` or `--updates` when manually running the Python service;
* `USER_FAILURES_DUMPS` (optional): the paths of csv/tsv files with the dump of the user profiles that failed to update the profile. The paths should be separated by `;`, it can also be set using the argument `-f` or `--failures` when manually running the Python service;
* `TIME_RANGE` (optional): the temporal range in which compute the analytics, example of allowed values are ["TODAY", "1D", "7D", "10D", "30D"], it can also be set using the argument `-r` or `--range` when manually running the Python service.

Alternatively to the `TIME_RANGE` arbitrary start and end time could be set using the following environment variables:

* `START_TIME`: the start time from which compute the analytics, must be in iso format, it can also be set using the argument `-s` or `--start` when manually running the Python service;
* `END_TIME`: the end time up to which compute the analytics, must be in iso format, it can also be set using the argument `-e` or `--end` when manually running the Python service.


#### Script profiles_and_tasks_m46

The script for getting the profiles and task in the newly requested format for M46 pilots. allows to set the following environment variables:

* `PROFILE_FILE`: the path of csv/tsv file where to store the profiles without personal information in the newly requested format for M46 pilots, it can also be set using the argument `-pf` or `--profile_file` when manually running the Python service;
* `TASK_FILE`: the path of csv/tsv file where to store the tasks in the newly requested format for M46 pilots, it can also be set using the argument `-tf` or `--task_file` when manually running the Python service;
* `INSTANCE` (optional, the default value is the development instance): the host of target instance, it can also be set using the argument `-i` or `--instance` when manually running the Python service;
* `APIKEY`: the apikey for accessing the services, it can also be set using the argument `-a` or `--apikey` when manually running the Python service;
* `APP_IDS`: the ids of the chatbots from which take the users. The ids should be separated by `;`, it can also be set using the argument `-ai` or `--app_ids` when manually running the Python service;
* `TIME_RANGE` (optional): the temporal range in which compute the analytics, example of allowed values are ["TODAY", "1D", "7D", "10D", "30D"], it can also be set using the argument `-r` or `--range` when manually running the Python service.

Alternatively to the `TIME_RANGE` arbitrary start and end time could be set using the following environment variables:

* `START_TIME`: the start time from which compute the analytics, must be in iso format, it can also be set using the argument `-s` or `--start` when manually running the Python service;
* `END_TIME`: the end time up to which compute the analytics, must be in iso format, it can also be set using the argument `-e` or `--end` when manually running the Python service.


## Usage

### Web service

This service can be run with the command:

```bash
python -m memex_logging.ws.main
```

### Script for computing the analytics

This service can be run with the command:

```bash
python -m memex_logging.compute_analytics_questions_users.main
```

### Script for getting the associations

This service can be run with the command:

```bash
python -m memex_logging.compute_analytics_questions_users.id_email
```

### Script for getting the apps usage

This service can be run with the command:

```bash
python -m memex_logging.compute_analytics_questions_users.apps_usage
```

### Script for getting the profiles and tasks (m46 format)

This service can be run with the command:

```bash
python -m memex_logging.compute_analytics_questions_users.profiles_and_tasks_m46
```


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Changelog](#changelog)
  - [Breaking changes from 1.* to 2.*](#breaking-changes-from-1-to-2)
  - [Version 2.*](#version-2)
    - [next](#next)
    - [2.3.0](#230)
    - [2.2.0](#220)
    - [2.1.2](#212)
    - [2.1.1](#211)
    - [2.1.0](#210)
    - [2.0.2](#202)
    - [2.0.1](#201)
    - [2.0.0](#200)
  - [Version 1.*](#version-1)
    - [1.4.0](#140)
    - [1.3.0](#130)
    - [1.2.1](#121)
    - [1.2.0](#120)
    - [1.1.0](#110)
    - [1.0.0](#100)
  - [Version 0.*](#version-0)
    - [0.2.0](#020)
    - [0.1.7](#017)
    - [0.1.6](#016)
    - [0.1.5](#015)
    - [0.1.4](#014)
    - [0.1.3](#013)
    - [0.1.2](#012)
    - [0.1.1](#011)
    - [0.1.0](#010)
    - [0.0.25](#0025)
    - [0.0.24](#0024)
    - [0.0.23](#0023)
    - [0.0.22](#0022)
    - [0.0.21](#0021)
    - [0.0.20](#0020)
    - [0.0.19](#0019)
    - [0.0.18](#0018)
    - [0.0.17](#0017)
    - [0.0.16](#0016)
    - [0.0.15](#0015)
    - [0.0.14](#0014)
    - [0.0.13](#0013)
    - [0.0.12](#0012)
    - [0.0.11](#0011)
    - [0.0.10](#0010)
    - [0.0.9](#009)
    - [0.0.8](#008)
    - [0.0.7](#007)
    - [0.0.6](#006)
    - [0.0.5](#005)
    - [0.0.4](#004)
    - [0.0.3](#003)
    - [0.0.2](#002)
    - [0.0.1](#001)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Changelog

## Breaking changes from 1.* to 2.* 

* the parameter `staticId` of the Analytic has been renamed to `id`
* the parameter `query` of the Analytic has been renamed to `descriptor`
* the key of the _segmentation result_ `count` has been renamed into `segments`
* The following metric values have been renamed 
  - `u:total` into `total`
  - `u:active` into `active`
  - `u:engaged` into `engaged`
  - `u:new` into `new`
  - `a:segmentation` into `age`
  - `g:segmentation` into `gender`
  - `m:from_users` into `requests`
  - `m:segmentation` into `all`
  - `r:segmentation` into `requests`
  - `u:segmentation` into `requests`
  - `m:from_bot` into `from_bot`
  - `m:responses` into `responses`
  - `m:notifications` into `notifications`
  - `t:total` into `total`
  - `t:active` into `active`
  - `t:closed` into `closed`
  - `t:new` into `new`
  - `t:segmentation` into `label`
  - `c:total` into `total`
  - `c:new` into `new`
  - `d:fallback` into `fallback`
  - `d:intents` into `intents`
  - `d:domains` into `domains`
  - `b:response` into `response`
* The keys `items` and `transactions` have been removed from the analytic result contents.

## Version 2.*

### 2.4.0

:rocket: New Features
* Added possibility to delete all the message for an user from the database.

:house: Internal
* Added user profiles extraction to the main script for extracting pilots stats.
* Added user profiles extraction without personal information to the app_usage script.
* Updated to Project template version `4.12.5`.
* Added script for the extraction of tasks and profiles in the newly requested format for M46 pilots.

### 2.3.0

* Integrated information about the number of asked questions and given answers by the users and also the first and last usage of the bot.
* Updated to Project template version `4.6.2`.
* Updated queries that use cardinality aggregation adding the precision threshold parameter.

### 2.2.0

* Updated queries used in count computations.
* Updated scripts for extracting stats adding also the information whether the users have enabled the survey app.
* Added script for extracting information on app usage by the users.
* Updated to Project template version `4.6.1`.

### 2.1.2

* Updated common models to version `3.1.0` in order to fix the representation errors due to the old Norm model used in the library.

### 2.1.1

* Solved an issue related to integration of endless time window when computing statistics for new users and new conversations.

### 2.1.0

* Added a new `all` analytic window type allowing to analyze all available data up to now.
* The support for periodical (currently the period is set to _once a day, at 4 a.m._) computation of not concluded _fixed time window_ analytic results have been added.
* Changed periodical re-computation of _moving time window_ analytic results execution time to _once a day, at 4 a.m._
* Solved an issue related to aggregations analytics changing aggregations results using always a dictionary.

### 2.0.2

* Message analytics have been updated.
  * Renamed type *from_user* into *requests*
  * Removed analytic of type *from_bot*
* Added two new task metrics: `new_active` and `new_closed`.
* Modified the `closed` task metric to get the number of tasks closed up to the end of a certain time range.
* Removed id of the application from indices of messages and analytics.
* Created an analytic dao for handling the interactions with the database.
* Set aggregation result to `None` if the field where we are performing the aggregation is not present on the message data.
* Added sentry integration in Migrator.

### 2.0.1

* Fixed to broken migration.

### 2.0.0

* Included timestamp details among the analytic result data.
* The support for periodical (currently the period is set to _once a day, at midnight_) re-computation of _moving time window_ analytic results have been added.
* The computation of analytic results is now not triggered automatically upon creation. In order for the results to be computed, it is necessary to wait for the periodical recomputation. It is also possible to take advantage of a new endpoint allowing to explicitly request the result computation.
* _Time window_ type values have been updated. In particular:
  * `DEFAULT` has bene renamed to `moving`
  * `CUSTOM` has been renamed to `fixed`
  Deprecated values are still supported at this stage during the creation step of analytics. Their support will be removed in the upcoming versions.

logger.py
from flask import Flask
from flask import request
# for second-level logging
import logging
from logging.handlers import RotatingFileHandler
import datetime
# for handling elasticsearch
from elasticsearch import Elasticsearch

import requests

import json

app = Flask(__name__)

# elasticsearch connection - params
ELASTIC_HOST = 'localhost'
ELASTIC_PORT = 9200

# second-level logging
handler = RotatingFileHandler('logger.log', maxBytes=10000, backupCount=1)
handler.setLevel(logging.INFO)
app.logger.addHandler(handler)

@app.errorhandler(400)
def internal_error(error):
    '''
    function that handle a 400 error and suggest possible solutions

    :param error: the error thrown
    :type error: string

    :output response: a response with status 400 and a checklist for a possible solution
    :type response: response
    '''
    response = app.response_class(
        response='400 CHECKLIST: data-format; data-limitations (i.e. mandatory vs suggested attributes)',
        status=400,
        mimetype='application/json'
    )
    app.logger.error('Logger @ UPPER 400 THROWN ' + str(datetime.datetime.now() ))
    return response

@app.errorhandler(404)
def not_found(error):
    '''
    function that handle a 404 error and suggest possible solutions

    :param error: the error thrown
    :type error: string

    :output response: a response with status 403 and a checklist for a possible solution
    :type response: response
    '''
    response = app.response_class(
        response='404 CHECKLIST: url and eventually uri',
        status=404,
        mimetype='application/json'
    )
    app.logger.error('Logger @ UPPER 404 THROWN ' + str(datetime.datetime.now() ))
    return response

@app.route('/messages', methods=['POST'])
def messages():
    '''
    to log an array of messages into the elasticsearch instance

    :param json: a valid json string in the body of the request
    :type json: string

    :output response: a response with a status and, eventually, a checklist
    :type response: response
    '''
    # second-level logging
    app.logger.warning('Logger @ start logging an array of messages - ' + str(datetime.datetime.now() ))
    # open a connection with elasticsearch
    es=Elasticsearch([{'host':ELASTIC_HOST,'port':ELASTIC_PORT}])
    # parse the message received
    messages_received = request.json
    # for each key in the dictionary, take the value (the message) and push it in the database
    # there should be just one key associated to a list of messages
    for k,v in messages_received.items():
        # if there is more than one top-level key
        if(len(k) > 1):
            response = app.response_class(
                response='400 CHECKLIST there must be just one top-level key in the json file (i.e. {messages:[{...},...]})',
                status=400,
                mimetype='application/json'
            )
            app.logger.warning('Logger @ wrong format of the message (tlkey) - ' + str(datetime.datetime.now() ))
            return response
        for element in v:
            # push the message in the database
            es.index(index='mem-ex',doc_type='message',body=element)
    # eventaully return with a 200 state
    response = app.response_class(
        response='Ok',
        status=200,
        mimetype='application/json'
    )
    app.logger.warning('Logger @ end logging an array of messages - ' + str(datetime.datetime.now() ))
    return response

@app.route('/message', methods=['POST'])
def message():
    '''
    to log an single message into the elasticsearch instance

    :param json: a valid json string in the body of the request
    :type json: string

    :output response: a response with a status and, eventually, a checklist
    :type response: response
    '''
    app.logger.warning('Logger @ start logging of a single message - ' + str(datetime.datetime.now() ))

    es=Elasticsearch([{'host':ELASTIC_HOST,'port':ELASTIC_PORT}])

    es.index(index='mem-ex',doc_type='message',body=request.json)

    response = app.response_class(
        response='Ok',
        status=200,
        mimetype='application/json'
    )
    app.logger.warning('Logger @ end logging of a single message - ' + str(datetime.datetime.now() ))
    return response
